runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,189 @@
+-- `nil` indicates following the global state.
+---@field enabled? boolean
+--
+-- Each client attached to this buffer must exists.
+--
+-- Index in the form of client_id -> (row -> highlights)
+---@field client_hilights table<integer, table<integer, lsp.DocumentHighlight[]?>?>
+
+-- Each buffer attached by at least one LSP server must exists,
+-- otherwise it should not exists or be cleaned up.
+--
+-- Index in the form of bufnr -> bufstate
+---@type table<integer, vim.lsp.document_highlight.BufState?>
+local bufstates = {}
+for _, client in ipairs(vim.lsp.get_clients()) do
+  for _, bufnr in ipairs(vim.lsp.get_buffers_by_client_id(client.id)) do
+    local bufstate = bufstates[bufnr] or {}
+    local client_hilights = bufstate.client_hilights or {}
+
+    if not client_hilights[client.id] then
+      client_hilights[client.id] = {}
+    end
+
+    bufstate.client_hilights = client_hilights
+    bufstates[bufnr] = bufstate
+  end
+end
+api.nvim_create_autocmd('LspAttach', {
+  group = augroup,
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.buf
+    local bufstate = bufstates[bufnr] or {}
+    bufstates[bufnr] = bufstate
+
+    ---@type integer
+    local client_id = args.data.client_id
+    local client_hilights = bufstate.client_hilights or {}
+    client_hilights[client_id] = {}
+    bufstate.client_hilights = client_hilights
+  end,
+})
+api.nvim_create_autocmd('LspDetach', {
+  group = augroup,
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.bufnr
+    local bufstate = assert(bufstates[bufnr])
+
+    ---@type integer
+    local client_id = args.data.client_id
+    bufstate.client_hilights[client_id] = nil
+    if vim.tbl_isempty(bufstate.client_hilights) then
+      bufstates[bufnr] = nil
+    end
+  end,
+})
+
+-- |lsp-handler| for the method `textDocument/documentHighlight`
+-- Store hints for a specific buffer and client
+---@param result? lsp.DocumentHighlight[]
+---@param ctx lsp.HandlerContext
+function M.on_document_highlight(err, result, ctx)
+  if err then
+    log.error('document highlight', err)
+  end
+
+  local bufnr = assert(ctx.bufnr)
+  local bufstate = assert(bufstates[bufnr])
+
+  local client_id = assert(ctx.client_id)
+  local client_hilights = bufstate.client_hilights
+
+  ---@type table<integer, lsp.DocumentHighlight[]?>
+  local row_highlights = {}
+  for _, highlight in pairs(result or {}) do
+    for row = highlight.range['start'].line, highlight.range['end'].line do
+      local highlights = row_highlights[row] or {}
+      highlights[#highlights + 1] = highlight
+      row_highlights[row] = highlights
+    end
+  end
+  client_hilights[client_id] = row_highlights
+  api.nvim__redraw({ buf = bufnr, valid = true })
+end
+
+---@param bufnr integer
+local function refresh(bufnr)
+  local params = util.make_position_params()
+  vim.lsp.buf_request(bufnr, ms.textDocument_documentHighlight, params)
+end
+
+api.nvim_create_autocmd('CursorMoved', {
```

Comment:
It is better to avoid CursorMoved

Reply:
Because `textDocument/documentHighlight` needs to get the cursor's position in the document (it is used for fuzzy matching the references of the symbol under the cursor), I don't know know to avoid it. Since LSP requests are sent asynchronously, and we are not handling its return value here, I think it shouldn't cause too much of a performance issue, but I intended to implement debouncing here. If there is a better way, please let me know.

Reply:
You  to send the current cursor position to the server. What happens when you move quickly like holding jk is that the refresh function keeps sending requests. no processing here.

Use a timer here and store the last request. When it is triggered again and still within the timeout, you need to cancel the previous request and start a new timer. When the callback function is called, you also need to determine whether the cursor position is the same as the previous cursor position. This is no different from CursorHold. Users can adjust it according to updatetime

Reply:
> there send the current cursor position to the server. What happens when you move quickly like holding jk is that the refresh function keeps sending requests. no processing here.

As mentioned earlier, this is exactly my plan (I haven't optimized the performance yet; it's currently just a rough implementation, but I intend to do so, just like the todo list above).

I'm not sure if `CursorHold` is a good idea, It shouldn't be used for this purpose because we do want to trigger it after the cursor is moved. The default `updatetime` is too long for debouncing, and it's a global option that shouldn't be changed for one specific purpose.

Later when this PR is ready to merge, I will request a review.

Reply:
Yes CursorHold is not a good idea here either. CursorMoved is also sensitive. At least you need to handle some of the behaviors I said.

I'm not sure if we need this. Plugins need a certain amount of space. Not all in core.

Reply:
I don't use a plugin for this. I have this in my config:

```lua
do -- textDocument/documentHighlight
  local method = 'textDocument/documentHighlight'

  api.nvim_create_autocmd({ 'FocusGained', 'WinEnter', 'BufEnter', 'CursorMoved', 'CursorHold', 'CursorHoldI' }, {
    callback = debounce(200, function(args)
      lsp.buf.clear_references()
      local bufnr = args.buf --- @type integer
      for _, client in ipairs(lsp.get_clients({ bufnr = bufnr })) do
        if client.supports_method(method, { bufnr = bufnr }) then
          local params = lsp.util.make_position_params()
          client.request(method, params, nil, bufnr)
        end
      end
    end)
  })

  api.nvim_create_autocmd({ 'FocusLost', 'WinLeave', 'BufLeave' }, {
    callback = lsp.buf.clear_references
  })
end
```

It's not perfect, but works well enough.

Reply:
Tangential, but it would make sense to refactor this (which already exists somewhere in the changetracking module) and expose this. This is exactly the sort of short but non-trivial snippet that having in our stdlib would be a significant QoL improvement for the ecosystem.

(Same with `memoize` and async, *ceterum censeo*.)

Reply:
I think `schedule` and `schedule_wrap` work well as separate functions as they are both used plenty and they have different return types (`nil` vs `function`). It's similar to how Lua has `coroutine.wrap` vs `coroutine.create` which return `function` and `thread` respectively.

Reply:
In this case, I believe the closest-to-perfect solution would require resolving https://github.com/neovim/neovim/issues/27509. Other methods seem more like workarounds.

Reply:
(Mark as resolved because the debounce logic has been added.)